funfact/lang/interpreter/_base.py

```diff
@@ -121,65 +131,90 @@ def wrapped_f(*args, **kwargs):
+    def literal(self, value: LiteralValue, **payload):
         pass
 
     @abstractmethod
-    def tensor(self, abstract: AbstractTensor, **payload: Any):
+    def tensor(self, abstract: AbstractTensor, **payload):
         pass
 
     @abstractmethod
-    def index(self, item: AbstractIndex, bound: bool, kron: bool,
-              **payload: Any):
+    def index(self, item: AbstractIndex, bound: bool, kron: bool, **payload):
         pass
 
     @abstractmethod
-    def indices(self, items: Tuple[AbstractIndex], **payload: Any):
+    def indices(self, items: Tuple[AbstractIndex], **payload):
         pass
 
     @abstractmethod
     def index_notation(
-        self, tensor: P.tensor, indices: P.indices, **payload: Any
+        self, indexless: P.Numeric, indices: P.indices, **payload
     ):
         pass
 
     @abstractmethod
-    def call(self, f: str, x: Tensorial, **payload: Any):
+    def call(self, f: str, x: P.Tensorial, **payload):
         pass
 
     @abstractmethod
-    def pow(self, base: Numeric, exponent: Numeric, **payload: Any):
+    def pow(self, base: P.Numeric, exponent: P.Numeric, **payload):
         pass
 
     @abstractmethod
-    def neg(self, x: Numeric, **payload: Any):
+    def neg(self, x: P.Numeric, **payload):
         pass
 
     @abstractmethod
-    def ein(self, lhs: Numeric, rhs: Numeric, precedence: int, reduction: str,
-            pairwise: str, outidx: Optional[P.indices], **payload: Any):
+    def binary(
+        self, lhs: P.Numeric, rhs: P.Numeric, precedence: int, pairwise: str,
+        **payload
+    ):
         pass
 
     @abstractmethod
-    def tran(self, src: Numeric, indices: P.indices):
+    def ein(
+        self, lhs: P.Numeric, rhs: P.Numeric, precedence: int, reduction: str,
+        pairwise: str, outidx: Optional[P.indices], **payload
+    ):
+        pass
+
+    @abstractmethod
+    def tran(self, src: P.Numeric, indices: P.indices):
         pass
 
     def __call__(self, node: _ASNode, parent: _ASNode = None):
         node = copy.copy(node)
-        for name, value in node.fields_fixed.items():
-            setattr(node, name, _deep_apply(self, value, node))
         rule = getattr(self, node.name)
-        payload = rule(**node.fields)
-        if isinstance(payload, dict):
-            node.__dict__.update(**payload)
-        elif isinstance(payload, list):
-            for key, value in payload:
-                setattr(node, key, value)
-        elif isinstance(payload, tuple) and len(payload) == 2:
-            setattr(node, *payload)
-        else:
-            raise TypeError(f'Unrecognizable type for payload {payload}')
+
+        def _do_node():
+            payload = rule(**node.fields)
+            self.emplace(node, payload)
+
+        def _do_children():
+            for name, value in node.fields_fixed.items():
+                setattr(node, name, _deep_apply(self, value, node))
+
+        if self._traversal_order == self.TraversalOrder.PRE:
```

Comment:
Great that you added both options in a clean way!

Reply:
:sunglasses: